feat(CI/nix): init docs on CI and add pre-commit hooks

DavHau opened the pull request:
### Summary
- Add a .md file for each existing github workflow .yml documenting the workflow
- Introduce [cachix/pre-commit-hooks.nix](https://github.com/cachix/pre-commit-hooks.nix)
- add pre-commit-hook to check the nix formatting
- add pre-commit-hook to ensure that all workflow .yml files are documented
- remove the `Unset variables: ...` prints from the pure shell script writer

@jost-s @steveeJ 


### TODO:
- [x] CHANGELOG(s) updated with appropriate info
- [ ] Just before pressing the merge button, ensure new entries to CHANGELOG(s) are still under the _UNRELEASED_ heading

steveeJ commented on .github/workflows/autoupdate.md:
```diff
@@ -0,0 +1,7 @@
+
+# TODO
+
+We already have `autorebase.yml`. Why is this still needed?
```
autoupdate creates merge commits while autorebase does a rebase

DavHau replied:
Any reason why we need both approaches? They seem to solve the same problem.

steveeJ replied:
functionally they're different and then it comes down to preference

jost-s replied:
?? Aren't both run on every push to develop?
And whose preference, I mean how does one get to choose?

DavHau replied:
Autoupdate specifies to only act if the `autoupdate:opt-in` label is set. Potentially there is another label to activate the autorebase. This should be documented somewhere and potentially be added to the PR template if we want contributors to make use of these.

steveeJ replied:
they always run but they only do something if the PR is labelled accordingly. the label names are `auto{rebase,update}:opt-in` respectively, and can be applied to a PR by the author or maintainers. one will rebase source onto target and the other will merge target into source.

steveeJ commented on .github/workflows/build.md:
```diff
@@ -0,0 +1,10 @@
+## build.yml
+
+Runs holochain tests:
```
i find it too early to document it in such detail, if at all. it requires manual updating of the markdown file every time tests are changed or relocated

steveeJ commented on .github/workflows/check_run_detect_release_pr_failure.md:
```diff
@@ -0,0 +1,8 @@
+# check_run_detect_release_pr_failure.yml
+
+Intentifies new branches created with `releas-` prefix and warns users about this in mattermost:
```
```suggestion
Intentifies new branches created with `release-` prefix and warns users about this in mattermost:
```

jost-s replied:
```suggestion
Identifies new branches created with `release-` prefix and warns users about this in mattermost:
```

steveeJ reviewed (commented):
on a high level i'm worried about creating maintenance burden with this. specifically, separating the doc file from the workflow file and documenting the workflows in such detail.

i would prefer if we added a header comment in the workflow files directly.

jost-s commented on .github/workflows/ssh-session.md:
```diff
@@ -0,0 +1,3 @@
+# ssh-session.yml
+
+An action triggered manually to debug the cached state for `releas.yml` and `release-prepare.yml`
```
```suggestion
An action triggered manually to debug the cached state for `release.yml` and `release-prepare.yml`
```

jost-s reviewed (changes requested):
I think these additions are helpful. I also prefer the docs to be moved to the workflow files for easier maintenance.

DavHau commented:
I moved the docs inside the .yml files as comments and applied the suggested fixes.
Concerning too much detail, I removed some for build.yml but kept it for release.yml and release-prepare.yml. I think for those it's currently harder to get an overview due to their sizes and therefore the bullet points are important.

Does this look fine for you @steveeJ ?

Also could you give ma a hint of why the `check_run_detect_release_pr_failure.yml` is needed, so I can add that to the docs?

steveeJ commented:
> Also could you give ma a hint of why the `check_run_detect_release_pr_failure.yml` is needed, so I can add that to the docs?

i don't think we need it anymore. it was used to detect failed PR checks that are part of a release. but we opted for pushing to github directly during a release.

DavHau commented:
I removed the `check_run_detect_release_pr_failure.yml` and added more comments to the autoupdate and autorebase workflows.

I think this is ready.

steveeJ commented on nix/modules/pre-commit.nix:
```diff
@@ -0,0 +1,7 @@
+{ self, lib, ... }: {
+  perSystem = { config, self', inputs', pkgs, ... }:
+    {
+      pre-commit.check.enable = true;
```
what kind of check is enabled here? is it just the one in the next line?

DavHau replied:
This makes the pre-commit checks become part of `nix flake check`, see: https://flake.parts/options/pre-commit-hooks-nix.html#opt-perSystem.pre-commit.check.enable

steveeJ commented on nix/modules/devShells.nix:
```diff
@@ -65,6 +65,7 @@
               export CARGO_CACHE_RUSTC_INFO=1
               export PATH="$CARGO_INSTALL_ROOT/bin:$PATH"
               export NIX_PATH="nixpkgs=${pkgs.path}"
+              ${config.pre-commit.installationScript}
```
the holonix devshell inherits from rustDev. i don't know if it's appropriate to make this the default for all holonix users so i suggest we move this to coreDev.

DavHau replied:
This makes the pre-commit checks become part of `nix flake check`, see: https://flake.parts/options/pre-commit-hooks-nix.html#opt-perSystem.pre-commit.check.enable

jost-s reviewed (approved):
Good stuff
